### PR TITLE
Commvault Plugin: execute backup

### PR DESCRIPTION
**Pull request by razi-ahmad1**

### CODE REVIEW

File: src/main/groovy/com/morpheusdata/commvault/backup/openstack/CommvaultOpenstackBackupTypeProvider.groovy [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+	/**
+	 * A list of {@link OptionType OptionTypes} for use in the backup create and edit forms.
+	 * @return a list of option types
+	 */
+	@Override
+	Collection<OptionType> getOptionTypes() {
+		return new ArrayList<OptionType>()
+	}
+
+	BackupTypeProvider getBackupTypeProvider() {
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheusdata/commvault/backup/openstack/CommvaultOpenstackBackupTypeProvider.groovy]: Lines 151 to 165 appear to be Veeam specific methods that have been copied over to commvault. Please remove if they are not being used. Clean up the other backup type providers as well.

File: src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupTypeProvider.groovy [hunk 2]
```diff
@@ -21,9 +20,15 @@ class CommvaultBackupTypeProvider extends AbstractBackupTypeProvider {
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupTypeProvider.groovy:20]: This should really be an abstract class as it isn't meant to be used directly.

File: src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupTypeProvider.groovy [hunk 3]
```diff
@@ -33,7 +38,8 @@ class CommvaultBackupTypeProvider extends AbstractBackupTypeProvider {
 	 */
 	@Override
 	String getCode() {
-		return "commvaultBackupTypeProvider"
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupTypeProvider.groovy:36]: After converting to an abstract class `getCode()` and `getName()` can be removed.

File: src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupTypeProvider.groovy [hunk 4]
```diff
@@ -152,7 +158,7 @@ class CommvaultBackupTypeProvider extends AbstractBackupTypeProvider {
 	@Override
 	CommvaultBackupExecutionProvider getExecutionProvider() {
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupTypeProvider.groovy]: After converting this class to an abstract class the methods `getExecutionProvider()` and `getRestoreProvider()` should be left out of this class to force the implementing class to implement the methods.

File: src/main/groovy/com/morpheusdata/commvault/backup/file/CommvaultFileBackupTypeProvider.groovy [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+	 * 		 	</li>
+	 * 		 	<li>
+	 * 		 	  	TEMP_EXTRACT -- determines the visibility of the restore to new option
+	 * 		 	</li>
+	 * 		</ul>
+	 *
+	 * @return
+	 */
+	@Override
+	String getRestoreNewMode() {
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheusdata/commvault/backup/file/CommvaultFileBackupTypeProvider.groovy:129]: This should be null, see the BackupTypeSeed in embedded.

File: src/main/groovy/com/morpheusdata/commvault/backup/openstack/CommvaultOpenstackBackupTypeProvider.groovy [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+	/**
+	 * Refresh the provider with the associated data in the external system.
+	 * @param authConfig  necessary connection and credentials to connect to the external provider
+	 * @param backupProvider an instance of the backup integration provider
+	 * @return a {@link ServiceResponse} object. A ServiceResponse with a success value of 'false' will indicate the
+	 * refresh process has failed and will halt any further backup creation processes in the core system.
+	 */
+	@Override
+	ServiceResponse refresh(Map authConfig, BackupProviderModel backupProviderModel) {
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheusdata/commvault/backup/openstack/CommvaultOpenstackBackupTypeProvider.groovy]: methods `refresh()` and `clean()` can just be implemented in the parent abstract class. Same goes for the file and vmware backup type providers.

File: src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupExecutionProvider.groovy [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+				rtn.success = true
+
+			}
+			logoutSession(authConfig)
+		}
+		return rtn
+	}
+
+	def logoutSession(Map authConfig) {
+		authConfig.token = authConfig.token ?: CommvaultBackupUtility.getToken(authConfig.apiUrl, authConfig.username, authConfig.password)?.token
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupExecutionProvider.groovy]: Remove line 418. If we do not have a token then there is no session to logout.

**Review by ddeyoung — CHANGES_REQUESTED** [review]

File: src/main/groovy/com/morpheusdata/commvault/backup/openstack/CommvaultOpenstackBackupTypeProvider.groovy (hunk 1 above)
Comment (razi-ahmad1) [reply src/main/groovy/com/morpheusdata/commvault/backup/openstack/CommvaultOpenstackBackupTypeProvider.groovy]: @ddeyoung changes are made as suggested

File: src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupTypeProvider.groovy (hunk 2 above)
Comment (razi-ahmad1) [reply src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupTypeProvider.groovy:20]: @ddeyoung changes are made as suggested

File: src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupTypeProvider.groovy (hunk 3 above)
Comment (razi-ahmad1) [reply src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupTypeProvider.groovy:36]: @ddeyoung changes are made as suggested

File: src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupTypeProvider.groovy (hunk 4 above)
Comment (razi-ahmad1) [reply src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupTypeProvider.groovy]: @ddeyoung changes are made as suggested

File: src/main/groovy/com/morpheusdata/commvault/backup/file/CommvaultFileBackupTypeProvider.groovy (hunk 5 above)
Comment (razi-ahmad1) [reply src/main/groovy/com/morpheusdata/commvault/backup/file/CommvaultFileBackupTypeProvider.groovy:129]: @ddeyoung changes are made as suggested

File: src/main/groovy/com/morpheusdata/commvault/backup/openstack/CommvaultOpenstackBackupTypeProvider.groovy (hunk 6 above)
Comment (razi-ahmad1) [reply src/main/groovy/com/morpheusdata/commvault/backup/openstack/CommvaultOpenstackBackupTypeProvider.groovy]: @ddeyoung changes are made as suggested

File: src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupExecutionProvider.groovy (hunk 7 above)
Comment (razi-ahmad1) [reply src/main/groovy/com/morpheusdata/commvault/backup/CommvaultBackupExecutionProvider.groovy]: @ddeyoung changes are made as suggested

**Review by ddeyoung — APPROVED** [review]

**Pull request merged**